### PR TITLE
fix rebuild sender receiver disposing instance used by other threads …

### PR DESCRIPTION
**Pull request by VaclavK**
…concurrently

so ultimately I saw over 10K of ObjectDisposedExceptions during Captain Hook perf test thrown by Peter pan using Publish

the flow was as follow
 - Azure started issuing service quota exceeded, which the code tried to rectify by rebuilding the sender, but other threads already entered send flow and the client checks internally for inner objects (connection etc.) being disposed, which prompted all those exceptions - **the key here** was **high failure rate** (at that time 100%) resulting in **high rebuild rate**

I am aware of the fact that the retry was done primarily for auth stuff but since I cannot (and I doubt anybody) can compile complete list of exceptions to cover by rebuilding the client I merely attempted to fix in general

the async reader writer lock gives multi reader/single write exclusivity in async mode and so rebuild can be done in isolation

I also saw the race condition when adapters are created when the code was checking for tryadd returning false (as it means adapter for the type was already added by competing thread) by disposing it but then returning it

any questions, shout

**I added** a test for this - since I would not be able to reliably repro a service quota I trigger the exception by supplying connection string w/o send claim , this throws the exception upon send and reproduces the problem when under load

### CODE REVIEW

File: src/Eshopworld.Messaging/Messenger.cs [hunk 1]
```diff
@@ -215,6 +215,7 @@ internal async Task<IServiceBusNamespace> GetRefreshedServiceBusNamespace()
                 if (!ServiceBusAdapters.TryAdd(typeof(T), adapter))
                 {
                     adapter.Dispose();
+                    adapter = null;
```
Comment (VaclavK) [inline src/Eshopworld.Messaging/Messenger.cs:218]: this instance would have been otherwise returned below since not null

**Comment by djfr** [comment]
`AsyncReaderWriterLock` uses locks internally, which will block. So these, under high degrees of parallelism will give you very high contention rates on singleton Messenger's (which we told everyone to do back in the day).

I would either process the exceptions as you see them, by not rebuilding the clients on certain exceptions, or put in a specific retry policy on Send|Publish to capture specifically `ObjectDisposedException`. If I was executing this, I would do the former.

**Comment by VaclavK** [comment]
@djfr there is a tiny synchronisation but the key piece is this - see the note

private Task<IDisposable> RequestReaderLockAsync(CancellationToken cancellationToken)
        {
            lock (_mutex)
            {
                // If the lock is available or in read mode and there are no waiting writers, upgradeable readers, or upgrading readers, take it immediately.
                if (_locksHeld >= 0 && _writerQueue.IsEmpty)
                {
                    ++_locksHeld;
                    return Task.FromResult<IDisposable>(new ReaderKey(this));
                }
                else
                {
                    // Wait for the lock to become available or cancellation.
                    return _readerQueue.Enqueue(_mutex, cancellationToken);
                }
            }
        }

**Comment by VaclavK** [comment]
Putting policy on object dispose is imho not a good approach as ultimately my goal is to prevent them in the first place and retrying when they happen under high other exceptions rate ( where rebuild is warranted) will not really likely solve them

I am not opposed to decrease frequency of rebuilding the client, and thus decreasing the write phase, but i do not really know the list of those exceptions, authentication and likely others which i cannot really determines now 

But ultimately i am not writing this for unlikely doom scenario of constant consistent rebuilds 

I should also note that this wraps not the messenger but the sender and reader which are per topic/queue so that Singleton messenger is not a problem and wraps their rebuild independent of each other

File: src/Tests/Eshopworld.Messaging.Tests/MessengerTopicTest.cs [hunk 2]
```diff
@@ -36,6 +36,41 @@ public async Task Test_SendCreatesTheTopic()
...
+            {
+                tasks.Add(Task.Run(()=> msn.Publish(new TestMessage())));
+            }
+
+            Task global=null; 
+            try
+            {
+                global = Task.WhenAll(tasks);
+                await global;
+                Assert.True(!global.IsFaulted);
```
Comment (arturek) [inline src/Tests/Eshopworld.Messaging.Tests/MessengerTopicTest.cs]: in which case this assertion can fail? Isn't await always throws an exception if the awaited task is faulted?

**Review by dguerin — APPROVED** [review]

**Review by arturek — APPROVED** [review]

**Comment by djfr** [comment]
> Putting policy on object dispose is imho not a good approach as ultimately my goal is to prevent them in the first place and retrying when they happen under high other exceptions rate ( where rebuild is warranted) will not really likely solve them
> 
> I am not opposed to decrease frequency of rebuilding the client, and thus decreasing the write phase, but i do not really know the list of those exceptions, authentication and likely others which i cannot really determines now
> 
> But ultimately i am not writing this for unlikely doom scenario of constant consistent rebuilds
> 
> I should also note that this wraps not the messenger but the sender and reader which are per topic/queue so that Singleton messenger is not a problem and wraps their rebuild independent of each other

So, let's step back on why exactly this is being done:
- We're using Standard SB namespaces
- We're being throttled by SB, which leads to an exception, which leads to a client rebuild

The fix is actually, do what I said initially _"I would either process the exceptions as you see them, by not rebuilding the clients on certain exceptions"_, which is the right approach applied anywhere, not just SB. We wouldn't rebuild `ComosClient` if we got a `429` response either ...

Combined with moving the SBs to premium and scale them accordingly for performance tests and for normal production load.

This fix is essentially introducing locks on each send|publish, which will _always_ lead to thread contention under high parallelism - thus it's a mitigation effort, not a solution to the problem.

**Comment by VaclavK** [comment]
I can defo see

- migration to premium sb or follow MS recommendation to spread them out to multiple standard instances
- not rebuilding client for every exception - still not clear what the list role be but that can be obviously built up over time

As for the lock effect, this will be measured but if it of going to be what I expect it to be, this and above would sound solid to me ( minimising sender rebuild frequency while  covering for unknown unknown) **and may be even removed when we e.g. cover all other bases, while in the meantime decreasing message loss risk** 


I will release this as preview for the per test

**Comment by djfr** [comment]
If you're going with this mitigation, instead of addressing the problem, then have a look at [AsyncLocal](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=netcore-3.1), if you store the clients in `AsyncLocal` wrappers (and you manage root / child interactions), you can safely recycle them in that async execution context only and because most of the heavylifting is done in the CLR, so even if the CLR is locking threads, those locks are of a much shorter duration then doing locks in .Net code.

**Comment by VaclavK** [comment]
Will have a look, cheers

**Comment by VaclavK** [comment]
messaging exceptions from AI for the past 7 days


  | System.ObjectDisposedException | 8,631 |  
  | Microsoft.Azure.ServiceBus.QuotaExceededException | 12,525 |  
  | System.InvalidOperationException | 41

disclaimer - this is the outcome exception list not per each retry

File: src/Eshopworld.Messaging/Eshopworld.Messaging.csproj [hunk 3]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     
-    <Version>2.1.0</Version>
-    <PackageReleaseNotes>Added ability to override event names using EswEventName attribute</PackageReleaseNotes>
+    <Version>2.1.1-preview</Version>
```
Comment (micszym) [inline src/Eshopworld.Messaging/Eshopworld.Messaging.csproj]: @VaclavK now that is has been tested, do you plan on going with the non-preview version?

File: src/Eshopworld.Messaging/QueueAdapter.cs [hunk 4]
```diff
@@ -22,6 +23,7 @@ internal sealed class QueueAdapter<T> : ServiceBusAdapter<T>
         internal readonly IQueue AzureQueue;
 
         internal MessageSender Sender;
+        internal AsyncReaderWriterLock SenderLock = new AsyncReaderWriterLock();
```
Comment (micszym) [inline src/Eshopworld.Messaging/QueueAdapter.cs]: Any reason why this isn't `readonly`?

File: src/Eshopworld.Messaging/TopicAdapter.cs [hunk 5]
```diff
@@ -24,6 +25,7 @@ internal sealed class TopicAdapter<T> : ServiceBusAdapter<T>
         internal TopicClient Sender;
         internal ISubscription AzureTopicSubscription;
         internal string SubscriptionName;
+        internal AsyncReaderWriterLock senderLock = new AsyncReaderWriterLock();
```
Comment (micszym) [inline src/Eshopworld.Messaging/TopicAdapter.cs]: What is the naming strategy? Other classes use `SenderLock` instead... Might be worth keeping them in sync, also for the other internal members in the class.

File: src/Eshopworld.Messaging/TopicAdapter.cs [hunk 6]
```diff
@@ -97,7 +99,19 @@ internal async Task Send([NotNull]T message)
...
                     try
                     {
-                        await Sender.SendAsync(qMessage).ConfigureAwait(false);
+                        using (await senderLock.ReaderLockAsync())
+                        {
+                            await Sender.SendAsync(qMessage).ConfigureAwait(false);
+                        }
+                    }
+                    /**
+                     * we want to minimize the client rebuild frequency and ideally the reverse the approach
```
Comment (micszym) [inline src/Eshopworld.Messaging/TopicAdapter.cs]: typos -> `**the** reverse the approach`

**Review by micszym — APPROVED** [review]

File: src/Eshopworld.Messaging/Eshopworld.Messaging.csproj (hunk 3 above)
Comment (VaclavK) [reply src/Eshopworld.Messaging/Eshopworld.Messaging.csproj]: yeah the approach was to remove the preview - the preview was only to facilitate the peter pan perf test - resolved now

**Review by tabman83 — APPROVED** [review]

**Pull request merged**